libraries/Microsoft.Bot.Builder.Dialogs/Prompts/CloudOAuthPrompt.cs

```diff
@@ -0,0 +1,137 @@
+
+            // Check for timeout
+            if (OAuthHelper.HasTimeoutExpired(dc))
+            {
+                return dc.EndDialogAsync(null, cancellationToken);
+            }
+
+            if (dc.Context.Activity.Type == ActivityTypes.Message)
+            {
+                return OnContinueWithMessageActivityAsync(dc, userTokenClient, _settings.ConnectionName, _validator, _settings.EndOnInvalidMessage, cancellationToken);
+            }
+            else
+            {
+                return OAuthHelper.OnContinueWithNonMessageActivityAsync(dc, userTokenClient, _settings.ConnectionName, cancellationToken);
+            }
+        }
+
+        private static async Task<DialogTurnResult> OnContinueWithMessageActivityAsync(DialogContext dc, UserTokenClient userTokenClient, string connectionName, PromptValidator<TokenResponse> validator, bool endOnInvalidMessage, CancellationToken cancellationToken)
+        {
+            var tokenResponse = await OAuthHelper.CreateTokenResponseFromMessageAsync(userTokenClient, connectionName, dc.Context.Activity, cancellationToken).ConfigureAwait(false);
+
+            // Call any custom validation we might have.
+            if (await CheckValidatorAsync(dc, validator, tokenResponse, cancellationToken).ConfigureAwait(false))
+            {
+                return await dc.EndDialogAsync(tokenResponse, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // If EndOnInvalidMessage is set, complete the prompt with no result.
+                if (endOnInvalidMessage)
+                {
+                    return await dc.EndDialogAsync(null, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // If this was a message Activity we can use the retry prompt
+                    var promptOptions = dc.ActiveDialog.State[OAuthHelper.PersistedOptions].CastTo<PromptOptions>();
+                    if (!dc.Context.Responded && promptOptions?.RetryPrompt != null)
+                    {
+                        await dc.Context.SendActivityAsync(promptOptions.RetryPrompt, cancellationToken).ConfigureAwait(false);
+                    }
+
+                    return EndOfTurn;
+                }
+            }
+        }
+
+        private static async Task<bool> CheckValidatorAsync(DialogContext dc, PromptValidator<TokenResponse> validator, TokenResponse tokenResponse, CancellationToken cancellationToken)
+        {
+            if (validator != null)
+            {
+                // Increment attempt count.
+                var promptState = dc.ActiveDialog.State[OAuthHelper.PersistedState].CastTo<IDictionary<string, object>>();
+                promptState["AttemptCount"] = Convert.ToInt32(promptState["AttemptCount"], CultureInfo.InvariantCulture) + 1;
```

Comment:
nit: do other pieces of code access this state entry? consider defining constant at least at the method level